src/h3lib/lib/cellsToMultiPoly.c

```diff
@@ -0,0 +1,619 @@
+#include <stdbool.h>
+#include <stdint.h>
+#include <stdlib.h>
+#include <string.h>
+
+#include "algos.h"
+#include "alloc.h"
+#include "area.h"
+#include "h3Assert.h"
+#include "h3api.h"
+
+// After rough search, 10 seems to minimize compute time for large sets
+#define HASH_TABLE_MULTIPLIER 10
+
+typedef struct Arc {
+    H3Index id;
+
+    bool isVisited;
+    bool isRemoved;
+
+    // For doubly-arced list of edges in loop.
+    struct Arc *next;
+    struct Arc *prev;
+
+    // For union-find datastructure
+    // https://en.wikipedia.org/wiki/Disjoint-set_data_structure
+    struct Arc *parent;
+    int64_t rank;
+} Arc;
+
+typedef struct {
+    int64_t numArcs;
+    Arc *arcs;
+
+    // hash buckets for fast edge/arc lookup
+    int64_t numBuckets;
+    Arc **buckets;
+} ArcSet;
+
+typedef struct {
+    H3Index root;
+    double area;
+
+    GeoLoop loop;
+} SortableLoop;
+
+typedef struct {
+    int64_t numLoops;
+    SortableLoop *sloops;
+} SortableLoopSet;
+
+typedef struct {
+    double outerArea;
+    GeoPolygon poly;
+} SortablePoly;
+
+static inline int cmp_SortableLoop(const void *pa, const void *pb) {
+    const SortableLoop *a = (const SortableLoop *)pa;
+    const SortableLoop *b = (const SortableLoop *)pb;
+
+    // first, sort on connected component
+    if (a->root < b->root) return -1;
+    if (a->root > b->root) return 1;
+
+    // second, sort on area of loops
+    if (a->area < b->area) return -1;
+    if (a->area > b->area) return 1;
+
+    return 0;  // same root and equal area
+}
+
+static inline int cmp_SortablePoly(const void *pa, const void *pb) {
+    const SortablePoly *a = (const SortablePoly *)pa;
+    const SortablePoly *b = (const SortablePoly *)pb;
+
+    double A = a->outerArea;
+    double B = b->outerArea;
+
+    // Sort by area of outer loop, in descending order.
+    // NOTE: Using comparison trick because it is hard to hit the equality
+    // branch for coverage.
+    return (B > A) - (B < A);
+}
+
+static inline int cmp_raw(const void *a, const void *b) {
```

Comment:
I would call this cmp_int64 or cmp_H3Index, as the size of what is being compared is important